Userland/Services/WindowServer/WindowManager.cpp

```diff
@@ -1977,9 +1975,56 @@ Gfx::IntRect WindowManager::maximized_window_rect(Window const& window, bool rel
         });
     }
 
-    constexpr int tasteful_space_above_maximized_window = 1;
-    rect.set_y(rect.y() + tasteful_space_above_maximized_window);
-    rect.set_height(rect.height() - tasteful_space_above_maximized_window);
+    if (tile_type == WindowTileType::Maximized) {
+        auto border_thickness = palette().window_border_thickness();
+        rect.inflate(border_thickness * 2, border_thickness * 2);
+    }
+
+    if (tile_type == WindowTileType::Left
+        || tile_type == WindowTileType::TopLeft
+        || tile_type == WindowTileType::BottomLeft) {
+        rect.set_width(rect.width() / 2);
+    }
+
+    if (tile_type == WindowTileType::Right
+        || tile_type == WindowTileType::TopRight
+        || tile_type == WindowTileType::BottomRight) {
+        rect.set_width(rect.width() / 2);
+        rect.set_x(rect.width());
+    }
+
+    if (tile_type == WindowTileType::Top
+        || tile_type == WindowTileType::TopLeft
+        || tile_type == WindowTileType::TopRight) {
+        rect.set_height(rect.height() / 2);
+    }
+
+    if (tile_type == WindowTileType::Bottom
+        || tile_type == WindowTileType::BottomLeft
+        || tile_type == WindowTileType::BottomRight) {
+        auto half_screen_reminder = rect.height() % 2;
+        rect.set_height(rect.height() / 2 + half_screen_reminder);
+        rect.set_y(rect.height() - half_screen_reminder);
+    }
+
+    Gfx::IntRect window_rect = window.rect();
+    Gfx::IntRect window_frame_rect = window.frame().rect();
+
+    if (tile_type == WindowTileType::VerticallyMaximized) {
+        rect.set_x(window_rect.x());
+        rect.set_width(window_rect.width());
+    } else {
+        rect.set_x(rect.x() + window_rect.x() - window_frame_rect.x());
+        rect.set_width(rect.width() - window_frame_rect.width() + window_rect.width());
+    }
+
+    if (tile_type == WindowTileType::HorizontallyMaximized) {
+        rect.set_y(window_rect.y());
+        rect.set_height(window_rect.height());
+    } else {
+        rect.set_y(rect.y() + window_rect.y() - window_frame_rect.y());
+        rect.set_height(rect.height() - window_frame_rect.height() + window_rect.height());
+    }
 
     if (relative_to_window_screen)
```

Comment:
@thankyouverycool By the way, do you know what does this piece of code do? I thought it's related to virtual screens but `screen.rect().location()` is always `[0, 0]` no matter on which virtual screen we are.